src/internal_modules/roc_core/mov_histogram.h

```diff
@@ -0,0 +1,123 @@
+namespace core {
+
+//! @brief A class that implements a rolling window moving histogram.
+//! The MovHistogram class maintains a histogram of values within a specified window
+//! length. It divides the range of values into a specified number of bins and updates the
+//! histogram as new values are added and old values are removed from the window.
+//! @tparam T The type of values to be histogrammed.
+
+template <typename T> class MovHistogram {
+public:
+    //! @brief Constructs a moving histogram.
+    //! @param arena Memory arena for dynamic allocations.
+    //! @param value_range_min The minimum value of the range to be histogrammed.
+    //! @param value_range_max The maximum value of the range to be histogrammed.
+    //! @param num_bins The number of bins in the histogram. Each bin represents a
+    //! subrange of the value range.
+    //! @param window_length The length of the moving window. Only values within this
+    //! window are considered in the histogram.
+
+    MovHistogram(IArena& arena,
+                 T value_range_min,
+                 T value_range_max,
+                 size_t num_bins,
+                 size_t window_length)
+        : value_range_min_(value_range_min)
+        , value_range_max_(value_range_max)
+        , num_bins_(num_bins)
+        , window_length_(window_length)
+        , ring_buffer_(arena, window_length)
+        , bins_(arena)
+        , valid_(false) {
+        if (num_bins == 0 || window_length == 0 || value_range_min >= value_range_max) {
+            roc_panic("mov histogram: number of bins and window length must be greater "
+                      "than 0 and value_range_min must be less than value_range_max");
+        }
+
+        bin_width_ = (value_range_max - value_range_min) / static_cast<T>(num_bins);
+
+        if (!ring_buffer_.is_valid() || !bins_.resize(num_bins)) {
+            return;
+        }
+
+        valid_ = true;
+    }
+
+    //! Check if the histogram is valid.
+    bool is_valid() const {
+        return valid_;
+    }
+
+    //! Add a value to the histogram.
+    void add_value(const T& value) {
+        T clamped_value = value;
+
+        if (clamped_value < value_range_min_) {
+            clamped_value = value_range_min_;
+        } else if (clamped_value > value_range_max_) {
+            clamped_value = value_range_max_;
+        }
+
+        if (ring_buffer_.size() == window_length_) {
+            T oldest_value = ring_buffer_.front();
+            ring_buffer_.pop_front();
+            size_t oldest_bin_index = get_bin_index_(oldest_value);
+            bins_[oldest_bin_index]--;
+        }
+
+        ring_buffer_.push_back(clamped_value);
+        size_t new_bin_index = get_bin_index_(clamped_value);
+        if (new_bin_index < num_bins_) {
+            bins_[new_bin_index]++;
+        }
+    }
+
+    //! Get the number of values in the given bin.
+    size_t get_bin_counter(size_t bin_index) const {
+        return bins_[bin_index];
```

Comment:
if `bin_index` > `num_bins_` we should panic here

Reply:
Actually `Array` already has bound checking.

Reply:
ah, OK